### ISSUE TITLE
Unable to create ublk target on Ubuntu 22.04 with kernel 6.5.0-21, liburing2.6

### ISSUE THREAD

**Issue by araghuna1** [spdk/spdk#3282]
# Sighting report

**> sudo build/bin/spdk_tgt &**
          [2024-02-28 15:07:33.571302] Starting SPDK v24.05-pre git sha1 82abe8c20 / DPDK 23.11.0 initialization...
          [2024-02-28 15:07:33.571497] [ DPDK EAL parameters: spdk_tgt --no-shconf -c 0x1 --huge-unlink --no-telemetry --log-level=lib.eal:6 --log-level=lib.cryptodev:5 --log-level=user1:6 --iova-mode=pa --base-virtaddr=0x200000000000 --match-allocations --file-prefix=spdk_pid361718 ]
          [2024-02-28 15:07:33.633478] app.c: 815:spdk_app_start: *NOTICE*: Total cores available: 1
          [2024-02-28 15:07:33.712729] reactor.c: 937:reactor_run: *NOTICE*: Reactor started on core 0

**> sudo -E scripts/rpc.py ublk_create_target**
          [2024-02-28 15:07:45.822094] ublk.c: 502:ublk_ctrl_cmd_get_features: *ERROR*: No available sqe in ctrl ring
          [2024-02-28 15:07:45.822205] ublk.c: 695:ublk_create_target: *ERROR*: Fail to open UBLK, error=No such file or directory
          [2024-02-28 15:07:45.822235] ublk_rpc.c:  51:rpc_ublk_create_target: *ERROR*: Can't create ublk target: No such file or directory
          request:
          {
            "method": "ublk_create_target",
            "req_id": 1
          }
          Got JSON-RPC error response
          response:
          {
            "code": -32603,
            "message": "No such file or directory"
          }

## Expected Behavior
ublk target should be created successfully

## Steps to Reproduce
1. since ublk needs 6.X kernel, Upgrade ubuntu22.04 (which has v5.15 by default) to 6.X kernel:
 `> sudo apt-get install --install-recommends linux-generic-hwe-22.04`
    reboot into new kernel
2.  spdk configure script looks for ublk_cmd.h in standard include location
  `sudo cp /usr/src/linux-headers-6.5.0-21-generic/include/uapi/linux/ublk_cmd.h /usr/include/linux/`
3. since ublk needs liburing2.2+,
    `> git clone https://github.com/axboe/liburing`
     `> cd liburing; ./configure; make; sudo make install`
     ;This installed liburing2.6 on my system
4.  get and build spdk
     `> git clone https://github.com/spdk/spdk.git`
      `> ./configure --with-ublk`
       `> make`
5. run spdk_tgt
    `> sudo scripts/setup.sh`
     `> build/bin/spdk_tgt &`
6. create ublk target
    `> sudo modprobe ublk_drv`
      ;confirmed presence of /dev/ublk-control
    `> sudo scripts/rpc.py ublk_create_target`

## Context (Environment including OS version, SPDK version, etc.)
Ubuntu 22.04.1 LTS
linux Kernel: 6.5.0-21
spdk: spdk repo main branch at commit id 82abe8c20260832fbbf2ec1cd9b3f8215f3946e8

**Comment by arsiesys** [comment]
Hi,
I got the same issue and submitted a fix for review which I am hoping is correct:
https://review.spdk.io/gerrit/c/spdk/spdk/+/22141

I am not 100% sure of me but.. for me it works now.

**Comment by araghuna1** [comment]
Thanks for looking at this issue & the attempted fix @arsiesys !
I tried this out on my system, but it did not address the issue. I still see the same error message when trying to create ublk target - "No available sqe in ctrl ring"

As a sanity test, I built and ran ubdsrv (https://github.com/ming1/ubdsrv.git). I am able to successfully create a ubd device and run the tests in that repo. This seems to indicate that other system dependencies (kernel module, liburing) are ok.

**Comment by arsiesys** [comment]
Oh, indeed. I didn’t carefully read the error.
I got the almost same environment (os, kernel) and the —with-ublk wasn’t enough to have it loaded in spdk_tgt (the rpc call were literally not loaded). 
The fix I submitted fixed that issue for me but not yours that is slightly different. 

Sorry for mistakenly read the error and assumed it was the same !

**Comment by jimharris** [comment]
> Hi, I got the same issue and submitted a fix for review which I am hoping is correct: https://review.spdk.io/gerrit/c/spdk/spdk/+/22141
> 
> I am not 100% sure of me but.. for me it works now.

This fix can't be the correct one - it is saying that ublk is a blockdev module when it is not.

It might be fixing some problem on your system, but it's not the correct one, and can't be upstreamed.

You should consider posting separate issue for this, it seems to be a different problem than one original submitter is reporting.

**Comment by jimharris** [comment]
> # Sighting report
> **> sudo build/bin/spdk_tgt &** [2024-02-28 15:07:33.571302] Starting SPDK v24.05-pre git sha1 [82abe8c](https://github.com/spdk/spdk/commit/82abe8c20260832fbbf2ec1cd9b3f8215f3946e8) / DPDK 23.11.0 initialization... [2024-02-28 15:07:33.571497] [ DPDK EAL parameters: spdk_tgt --no-shconf -c 0x1 --huge-unlink --no-telemetry --log-level=lib.eal:6 --log-level=lib.cryptodev:5 --log-level=user1:6 --iova-mode=pa --base-virtaddr=0x200000000000 --match-allocations --file-prefix=spdk_pid361718 ] [2024-02-28 15:07:33.633478] app.c: 815:spdk_app_start: _NOTICE_: Total cores available: 1 [2024-02-28 15:07:33.712729] reactor.c: 937:reactor_run: _NOTICE_: Reactor started on core 0
> 
> **> sudo -E scripts/rpc.py ublk_create_target** [2024-02-28 15:07:45.822094] ublk.c: 502:ublk_ctrl_cmd_get_features: _ERROR_: No available sqe in ctrl ring [2024-02-28 15:07:45.822205] ublk.c: 695:ublk_create_target: _ERROR_: Fail to open UBLK, error=No such file or directory [2024-02-28 15:07:45.822235] ublk_rpc.c: 51:rpc_ublk_create_target: _ERROR_: Can't create ublk target: No such file or directory request: { "method": "ublk_create_target", "req_id": 1 } Got JSON-RPC error response response: { "code": -32603, "message": "No such file or directory" }
> 
> ## Expected Behavior
> ublk target should be created successfully
> 
> ## Steps to Reproduce
> 1. since ublk needs 6.X kernel, Upgrade ubuntu22.04 (which has v5.15 by default) to 6.X kernel:
>    `> sudo apt-get install --install-recommends linux-generic-hwe-22.04`
>    reboot into new kernel
> 2. spdk configure script looks for ublk_cmd.h in standard include location
>    `sudo cp /usr/src/linux-headers-6.5.0-21-generic/include/uapi/linux/ublk_cmd.h /usr/include/linux/`
> 3. since ublk needs liburing2.2+,
>    `> git clone https://github.com/axboe/liburing`
>    `> cd liburing; ./configure; make; sudo make install`
>    ;This installed liburing2.6 on my system
> 4. get and build spdk
>    `> git clone https://github.com/spdk/spdk.git`
>    `> ./configure --with-ublk`
>    `> make`
> 5. run spdk_tgt
>    `> sudo scripts/setup.sh`
>    `> build/bin/spdk_tgt &`
> 6. create ublk target
>    `> sudo modprobe ublk_drv`
>    ;confirmed presence of /dev/ublk-control
>    `> sudo scripts/rpc.py ublk_create_target`
> 
> ## Context (Environment including OS version, SPDK version, etc.)
> Ubuntu 22.04.1 LTS linux Kernel: 6.5.0-21 spdk: spdk repo main branch at commit id [82abe8c](https://github.com/spdk/spdk/commit/82abe8c20260832fbbf2ec1cd9b3f8215f3946e8)

What is value of /sys/module/ublk_drv/parameters/ublks_max on your system?

**Comment by araghuna1** [comment]
> cat /sys/module/ublk_drv/parameters/ublks_max
64

**Comment by changpe1** [comment]
@araghuna1 I didn't try the same kernel version as yours but have a try on 6.5.6-300.fc39.x86_64, it works well, you may have a try with other kernel versions.

**Comment by araghuna1** [comment]
I have managed to successfully create ublk target ! The issue on my system was that there was another version of liburing (liburing.so.2.1.0) in /usr/lib/x86_64-linux-gnu/
I removed it by running `> sudo apt-get purge liburing2`
Then I again ran `> ./configure --with-ublk` and rebuilt. 
Now when I run the spdk_tgt application, everything works as expected.

So, while the spdk configure script check for liburing seemed to look for liburing in /usr/lib (it was failing without the liburing.so.xx in this directory), I'm guessing somewhere else in the script, /usr/lib/x86_64-linux-gnu directory got added to the LD_LIBRARY_PATH, causing the liburing.so.xx in /usr/lib/x86_64-linux-gnu to get linked, instead of /usr/lib/liburing.so.xx. 
_(The ubdsrv codebase on the other hand, found & linked the liburing.so.xx from /usr/lib. Hence it worked fine while the spdk_tgt app did not)_

My learning here -- look for liburing.so.* in /usr/lib, /usr/lib/x86_64-linux-gnu/ , /usr/lib64 etc. Ensure that there are no older versions around which might inadvertently get used, instead of the liburing version you wish to link to!

Thanks @arsiesys @jimharris @changpe1 for looking into this issue !
